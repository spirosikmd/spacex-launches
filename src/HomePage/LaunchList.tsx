import React from 'react';
import PropTypes from 'prop-types';
import {
  withStyles,
  Theme,
  WithStyles,
  createStyles,
} from '@material-ui/core/styles';
import Launch, { LaunchPropType } from './Launch';
import Image from '../Image';
import emptyListImage from './emptyList.png';
import { LaunchData } from '../api';

const styles = (theme: Theme) =>
  createStyles({
    launchList: {
      display: 'flex',
      flexDirection: 'column' as 'column',
      position: 'relative' as 'relative',
    },
    timeline: {
      borderLeft: `1px solid ${theme.palette.grey[300]}`,
      position: 'absolute' as 'absolute',
      top: '6px',
      bottom: '6px',
      left: `${theme.spacing.unit + 4}px`,
      '&::before': {
        position: 'absolute' as 'absolute',
        content: '""',
        borderLeft: `1px dashed ${theme.palette.grey[300]}`,
        top: '-13px',
        left: '-1px',
        height: '12px',
      },
      '&::after': {
        position: 'absolute' as 'absolute',
        content: '""',
        borderLeft: `1px dashed ${theme.palette.grey[300]}`,
        bottom: '-13px',
        left: '-1px',
        height: '12px',
      },
    },
    launchItem: {
      padding: `${theme.spacing.unit * 2}px 0`,
    },
  });

interface LaunchListProps extends WithStyles<typeof styles> {
  launches: LaunchData[];
}

const LaunchList = ({ classes, launches }: LaunchListProps) => {
  if (launches.length === 0) {
    return <Image src={emptyListImage} alt="No launches..." />;
  }

  return (
    <div className={classes.launchList}>
      <div className={classes.timeline} />
      {launches.map(launch => (
        <div key={launch.flightNumber} className={classes.launchItem}>
          <Launch launch={launch} />
        </div>
      ))}
    </div>
  );
};

LaunchList.propTypes = {
  classes: PropTypes.object.isRequired,
  launches: PropTypes.arrayOf(PropTypes.shape(LaunchPropType)).isRequired,
};

export default withStyles(styles)(React.memo(LaunchList));
