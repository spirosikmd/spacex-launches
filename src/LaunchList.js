import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import Launch, { LaunchPropType } from './Launch';
import emptyListImage from './emptyList.png';

const styles = theme => ({
  launchList: {
    display: 'flex',
    flexDirection: 'column',
    position: 'relative',
  },
  timeline: {
    borderLeft: `1px solid ${theme.palette.grey[300]}`,
    position: 'absolute',
    top: '6px',
    bottom: '6px',
    left: `${theme.spacing.unit + 4}px`,
    '&::before': {
      position: 'absolute',
      content: '""',
      borderLeft: `1px dashed ${theme.palette.grey[300]}`,
      top: '-13px',
      left: '-1px',
      height: '12px',
    },
    '&::after': {
      position: 'absolute',
      content: '""',
      borderLeft: `1px dashed ${theme.palette.grey[300]}`,
      bottom: '-13px',
      left: '-1px',
      height: '12px',
    },
  },
  launchItem: {
    padding: `${theme.spacing.unit * 2}px 0`,
  },
  emptyListImageWrapper: {
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
  },
  emptyListImage: {
    width: '400px',
    [theme.breakpoints.down('xs')]: {
      width: '300px',
    },
  },
});

export const LaunchList = ({ classes, launches }) => {
  if (launches.length === 0) {
    return (
      <div className={classes.emptyListImageWrapper}>
        <img
          src={emptyListImage}
          alt="No launches..."
          className={classes.emptyListImage}
        />
      </div>
    );
  }

  return (
    <div className={classes.launchList}>
      <div className={classes.timeline} />
      {launches.map(launch => (
        <div key={launch.flightNumber} className={classes.launchItem}>
          <Launch launch={launch} />
        </div>
      ))}
    </div>
  );
};

LaunchList.propTypes = {
  classes: PropTypes.object.isRequired,
  launches: PropTypes.arrayOf(PropTypes.shape(LaunchPropType)).isRequired,
};

export default withStyles(styles)(React.memo(LaunchList));
