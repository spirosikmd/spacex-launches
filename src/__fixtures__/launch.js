export function createLaunchResponse(data) {
  return {
    flight_number: 1,
    mission_name: 'Test',
    launch_date_utc: '2019-01-01T18:16:00.000Z',
    launch_success: true,
    upcoming: false,
    is_tentative: false,
    details: 'Details',
    mission_id: ['id1'],
    links: {
      mission_patch: 'missionPatch',
    },
    ...data,
  };
}

export function createLaunch(data) {
  return {
    flightNumber: 1,
    missionName: 'Test',
    utcDate: new Date('2019-01-01'),
    isSuccessful: true,
    isFailed: false,
    isUpcoming: false,
    isInProgress: false,
    isTentative: false,
    details: 'Details',
    missionIds: ['id1'],
    missionPatch: 'missionPatch',
    ...data,
  };
}
