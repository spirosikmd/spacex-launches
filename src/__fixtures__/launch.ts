import { LaunchResponseData, LaunchData } from '../api';

export function createLaunchResponse(
  data: LaunchResponseData
): LaunchResponseData {
  return {
    flight_number: 1,
    mission_name: 'Test',
    launch_date_utc: '2019-01-01T18:16:00.000Z',
    launch_success: true,
    upcoming: false,
    is_tentative: false,
    details: 'Details',
    mission_id: ['id1'],
    links: {
      mission_patch: 'missionPatch',
    },
    ...data,
  };
}

export function createLaunch(data: LaunchData): LaunchData {
  return {
    flightNumber: 1,
    missionName: 'Test',
    utcDate: new Date('2019-01-01'),
    isSuccessful: true,
    isFailed: false,
    isUpcoming: false,
    isInProgress: false,
    isTentative: false,
    details: 'Details',
    missionIds: ['id1'],
    missionPatch: 'missionPatch',
    ...data,
  };
}
