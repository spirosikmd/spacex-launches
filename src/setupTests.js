import React from 'react';
import { configure, shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { createSerializer } from 'enzyme-to-json';
import { unwrap } from '@material-ui/core/test-utils';

configure({ adapter: new Adapter() });

expect.addSnapshotSerializer(createSerializer({ mode: 'deep' }));

global.shallow = (Component, props = {}) => {
  const ComponentWithoutStyles = unwrap(Component);
  return shallow(<ComponentWithoutStyles {...props} />);
};

global.mount = (Component, props = {}) => {
  const ComponentWithoutStyles = unwrap(Component);
  return mount(<ComponentWithoutStyles {...props} />);
};

global.Date = jest.fn().mockImplementation(() => {
  return {
    toUTCString: jest.fn().mockReturnValue('Fri, 11 Jan 2019 15:31:00 GMT'),
    getDate: jest.fn().mockReturnValue(3),
    getMonth: jest.fn().mockReturnValue(10),
    getFullYear: jest.fn().mockReturnValue(2019),
  };
});

// https://github.com/airbnb/enzyme/issues/1875
jest.mock('react', () => {
  const r = jest.requireActual('react');
  return { ...r, memo: x => x };
});
