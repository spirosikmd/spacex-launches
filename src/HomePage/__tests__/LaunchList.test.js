import React from 'react';
import { render } from '../../setupTests';
import LaunchList from '../LaunchList';
import { createLaunch } from '../../__fixtures__/launch';

describe('LaunchList', () => {
  let props;

  beforeEach(() => {
    props = {
      launches: [],
      classes: {
        launchList: 'launchList',
        timeline: 'timeline',
        launchItem: 'launchItem',
      },
    };
  });

  describe('when there are no launches', () => {
    it('renders a message', () => {
      const { asFragment } = render(<LaunchList {...props} />);
      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('when there are launches', () => {
    beforeEach(() => {
      props.launches = [createLaunch(), createLaunch({ flightNumber: 2 })];
    });

    it('renders a list of launches', () => {
      const { asFragment } = render(<LaunchList {...props} />);
      expect(asFragment()).toMatchSnapshot();
    });
  });
});
