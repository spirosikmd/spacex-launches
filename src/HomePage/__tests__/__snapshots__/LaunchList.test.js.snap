// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LaunchList when there are launches renders a list of launches 1`] = `
<DocumentFragment>
  <div
    class="LaunchList-launchList-6 launchList"
  >
    <div
      class="LaunchList-timeline-7 timeline"
    />
    <div
      class="LaunchList-launchItem-8 launchItem"
    >
      <div
        class="Launch-launch-9"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation1 LaunchStatus-status-12 LaunchStatus-success-15 MuiPaper-rounded"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root LaunchStatus-statusIcon-13"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M0 0h24v24H0V0z"
              fill="none"
            />
            <path
              d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zM8 17.5c-1.38 0-2.5-1.12-2.5-2.5s1.12-2.5 2.5-2.5 2.5 1.12 2.5 2.5-1.12 2.5-2.5 2.5zM9.5 8c0-1.38 1.12-2.5 2.5-2.5s2.5 1.12 2.5 2.5-1.12 2.5-2.5 2.5S9.5 9.38 9.5 8zm6.5 9.5c-1.38 0-2.5-1.12-2.5-2.5s1.12-2.5 2.5-2.5 2.5 1.12 2.5 2.5-1.12 2.5-2.5 2.5z"
            />
          </svg>
        </div>
        <div
          class="Launch-connector-11"
        />
        <div
          class="MuiPaper-root MuiPaper-elevation1 MuiCard-root Launch-launchInfo-10 MuiPaper-rounded"
        >
          <div
            class="MuiCardHeader-root"
          >
            <div
              class="MuiCardHeader-content"
            >
              <span
                class="MuiTypography-root MuiCardHeader-title MuiTypography-h5 MuiTypography-displayBlock"
              >
                Test
              </span>
              <span
                class="MuiTypography-root MuiCardHeader-subheader MuiTypography-body1 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
              >
                <span>
                  Tue, 01 Jan 2019 00:00:00 GMT
                </span>
              </span>
            </div>
          </div>
          <div
            class="MuiCardContent-root"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom"
            >
              Details
            </p>
            <p
              class="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary"
            >
              ID: id1
            </p>
          </div>
          <div
            class="MuiCardActions-root MuiCardActions-spacing"
          >
            <a
              aria-disabled="false"
              class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeSmall"
              href="/launches/1"
              role="button"
              tabindex="0"
            >
              <span
                class="MuiButton-label"
              >
                explore
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </a>
          </div>
        </div>
      </div>
    </div>
    <div
      class="LaunchList-launchItem-8 launchItem"
    >
      <div
        class="Launch-launch-9"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation1 LaunchStatus-status-12 LaunchStatus-success-15 MuiPaper-rounded"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root LaunchStatus-statusIcon-13"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M0 0h24v24H0V0z"
              fill="none"
            />
            <path
              d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zM8 17.5c-1.38 0-2.5-1.12-2.5-2.5s1.12-2.5 2.5-2.5 2.5 1.12 2.5 2.5-1.12 2.5-2.5 2.5zM9.5 8c0-1.38 1.12-2.5 2.5-2.5s2.5 1.12 2.5 2.5-1.12 2.5-2.5 2.5S9.5 9.38 9.5 8zm6.5 9.5c-1.38 0-2.5-1.12-2.5-2.5s1.12-2.5 2.5-2.5 2.5 1.12 2.5 2.5-1.12 2.5-2.5 2.5z"
            />
          </svg>
        </div>
        <div
          class="Launch-connector-11"
        />
        <div
          class="MuiPaper-root MuiPaper-elevation1 MuiCard-root Launch-launchInfo-10 MuiPaper-rounded"
        >
          <div
            class="MuiCardHeader-root"
          >
            <div
              class="MuiCardHeader-content"
            >
              <span
                class="MuiTypography-root MuiCardHeader-title MuiTypography-h5 MuiTypography-displayBlock"
              >
                Test
              </span>
              <span
                class="MuiTypography-root MuiCardHeader-subheader MuiTypography-body1 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
              >
                <span>
                  Tue, 01 Jan 2019 00:00:00 GMT
                </span>
              </span>
            </div>
          </div>
          <div
            class="MuiCardContent-root"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom"
            >
              Details
            </p>
            <p
              class="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary"
            >
              ID: id1
            </p>
          </div>
          <div
            class="MuiCardActions-root MuiCardActions-spacing"
          >
            <a
              aria-disabled="false"
              class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeSmall"
              href="/launches/2"
              role="button"
              tabindex="0"
            >
              <span
                class="MuiButton-label"
              >
                explore
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`LaunchList when there are no launches renders a message 1`] = `
<DocumentFragment>
  <div
    class="Image-imageWrapper-4"
  >
    <img
      alt="No launches..."
      class="Image-image-5"
      src="emptyList.png"
    />
  </div>
</DocumentFragment>
`;
